### ISSUE TITLE
[REVIEW]: Asimov: A framework for coordinating parameter estimation workflows

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4170]
**Submitting author:** @transientlunatic (<a href="http://orcid.org/0000-0003-3772-198X">Daniel Williams</a>)
**Repository:** <a href="https://git.ligo.org/asimov/asimov" target ="_blank">https://git.ligo.org/asimov/asimov</a>
**Version:** v0.3.3
**Editor:** @mbobra
**Reviewer:** @danehkar, @matthewfeickert
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199/status.svg)](https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199"><img src="https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199/status.svg)](https://joss.theoj.org/papers/ef5eb6f46867ddf5b10667a8357cf199)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://git.ligo.org/asimov/asimov" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@danehkar & @matthewfeickert, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mbobra know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @danehkar
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://git.ligo.org/asimov/asimov">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://git.ligo.org/asimov/asimov">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@transientlunatic) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @matthewfeickert
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://git.ligo.org/asimov/asimov">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://git.ligo.org/asimov/asimov">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@transientlunatic) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @danehkar, @matthewfeickert it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.58 s (173.7 files/s, 21144.6 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          50           1361           1875           4585
INI                             10            197              7           1098
reStructuredText                30            705            673            993
YAML                             3             15             26            199
XML                              1              0              0            179
SVG                              1              1              1             76
Markdown                         2             46              0             67
DOS Batch                        1              8              1             26
make                             1              4              7              9
Dockerfile                       1              6              2              5
-------------------------------------------------------------------------------
SUM:                           100           2343           2592           7237
-------------------------------------------------------------------------------


Statistical information for the repository '6b738369118957b8468d3b0f' was
gathered on 2022/02/16.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Avi Vajpeyi                      1            70              1            0.44
Daniel Williams                232         12013           3961           98.30
James Alexander Clar             1           120              0            0.74
Richard Philip Udall             3            81              5            0.53

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Avi Vajpeyi                  56           80.0         10.4               12.50
Daniel Williams            7615           63.4          9.2                9.93
James Alexander Clar         97           80.8         11.3                5.15
Richard Philip Udall         53           65.4          9.7                5.66
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4170 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by mbobra** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4170 with the following error: 
```
 
```

**Comment by mbobra** [comment]
@Kevin-Mattheus-Moerman Any idea why the pdf won't compile?

**Comment by danielskatz** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by danielskatz** [comment]
@mbobra - maybe just a transient error...

**Comment by danielskatz** [comment]
@whedon check references from branch joss-paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1103/PhysRevX.11.021053 is OK
- 10.1007/s41114-020-00026-9 is OK
- 10.1103/PhysRevD.91.042003 is OK
- 10.3847/1538-4365/ab06fc is OK
- 10.1103/PhysRevD.101.024056 is OK
- 10.1103/PhysRevD.102.044055 is OK
- 10.1103/PhysRevLett.116.241102 is OK
- 10.1103/PhysRevLett.116.061102 is OK
- 10.1016/j.softx.2021.100765 is OK
- 10.1103/PhysRevD.103.104056 is OK
- 10.5281/zenodo.4024432 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mbobra** [comment]
Ah thank you @danielskatz !

**Comment by mbobra** [comment]
👋 @danehkar @matthewfeickert Thank you so much for agreeing to review! You can find the article above ⬆️ , and the checklist and software repository linked in the first comment box on this issue. I think you're good to go. Again, JOSS is an open review process and we encourage communication between the reviewers, the submitting author, and the editor. So please feel free to ask questions at any time, I'm always around.

Can you please respond here (or give a thumbs up) so I know you're in the right place and found all the materials?

**Comment by matthewfeickert** [comment]
> Can you please respond here (or give a thumbs up) so I know you're in the right place and found all the materials?

All good to go @mbobra. :+1: Thanks for the info and the introductions. :slightly_smiling_face: 

@transientlunatic just to give you a heads up I'm finishing another JOSS review at the moment (will be done before end of the week for sure though) so I probably won't start on this review until the start of next week at the earliest.

**Comment by danehkar** [comment]
@mbobra :+1: Thanks for the information.

**Comment by matthewfeickert** [comment]
@transientlunatic @mbobra there's a problem that is going to greatly encumber the review. While the Asimov repo https://git.ligo.org/asimov/asimov is public, it is hosted on a LIGO/Virgo/KAGRA collaboration instance of GitLab that requires login access to open up Issues (which is normal and fine unto itself). The problem though is that as @danehkar (I think?) and I aren't members of the LIGO/Virgo/KAGRA collaboration we won't be able to open up any Issues for our comments in the review.

Can you suggest a way to proceed?

**Comment by transientlunatic** [comment]
Thanks very much for taking this on @danehkar, @matthewfeickert. 
I think the most straight-forward thing I can do is to set up a mirror of the repository on Github; I suspect trying to get you a login for our collaboration gitlab would be difficult, and inconvenient for yourselves, in addition to not being especially conducive to collaboration outwith the LVK. 
I'll let you know once I've set that up!

**Comment by transientlunatic** [comment]
I've now mirrored our collaboration repository on Github here:
https://github.com/ligo-asimov/asimov
Apologies for not having realised this would be a problem sooner!

**Comment by transientlunatic** [comment]
Hi @danehkar I think most of these things are already in-place for the package.
Do you require that unit tests are run via appveyor and travis? We already run these tests through gitlab CI.
We also maintain documentation for the master branch here: https://asimov.docs.ligo.org/asimov/
(I'm pushing a merge request through our internal review process to allow these to be built-out for all versions of the docs which will probably be of more use to this review).

**Comment by danehkar** [comment]
> We also maintain documentation for the master branch here: https://asimov.docs.ligo.org/asimov/

@transientlunatic It is OK. Please include its link in Readme file.

> The package is already available on conda forge.

Please also include the installation instruction for conda forge.

> Do you require that unit tests are run via appveyor and travis? We already run these tests through gitlab CI.

It is good to see that the package also runs with appveyor and travis without any issues.

**Comment by matthewfeickert** [comment]
> It is good to see that the package also runs with appveyor and travis without any issues.

I would recommend not spending your time with Travis unless you want have a dedicated amount of money that you're willing to pay into the future. Travis CI now (2020 onwards) [offers a open source projects a one time non-renewable 10,000 build credit](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing). This is not a sustainable model for open source projects unless there you have a dedicated fund for CI/CD.

In my mind, as long as you are testing the software on every OS and Python version that you **say that you support** this is sufficient testing. I haven't done my review yet, so maybe this is clear in the docs somewhere, but from the PyPI metadata classifiers

https://github.com/ligo-asimov/asimov/blob/b64d9f804e9e4ce7056ba4899fe3c9fb46304e44/setup.py#L39-L48

there is no OS information given, and while you're [testing against Python 3.6 through Python 3.9](https://github.com/ligo-asimov/asimov/blob/b64d9f804e9e4ce7056ba4899fe3c9fb46304e44/.gitlab-ci.yml#L8-L54) (metadata should get updated to reflect this and what about Python 3.10?) there is no testing against macOS and Windows. Azure Pipelines is a good choice here in my experience if you're not already using GitHub Actions (probably not the easiest with a mirror from GitLab) and if you don't already have a workflow you like with CricleCI. I am not saying Appveyor is bad, just that I prefer Azure Pipelines and in my experience it is a nicer UX.

**Comment by danehkar** [comment]
@matthewfeickert @transientlunatic 

> I would recommend not spending your time with Travis unless you want have a dedicated amount of money that you're willing to pay into the future. Travis CI now (2020 onwards) [offers a open source projects a one time non-renewable 10,000 build credit](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing). This is not a sustainable model for open source projects unless there you have a dedicated fund for CI/CD.

Travis is free for open source projects hosted on GitHub. If the package does not have any issues, it will take less 20 minutes to setup its test account on Travis. I hosted test units for several Python packages on Travis without spending any money. Each time, I update a code in a package, it automatically runs a test on my package and shows the passing status if there is no error. You need to have [.travis.yml](https://github.com/equib/pyEQUIB/blob/master/.travis.yml) in your repository, and a test folder, e.g., [pyequibtests](https://github.com/equib/pyEQUIB/tree/master/pyequibtests). To make an account on Travis, you need to use your GitHub account.

**Comment by matthewfeickert** [comment]
> Travis is free for open source projects hosted on GitHub.

Unfortunately this is no longer the case (though is was for many years). Please read what I shared. Though it is easy enough to see this for yourself.

If you visit travis-ci.org you'll be greeted with

> Since June 15th, 2021, the building on [travis-ci.org](http://www.travis-ci.org/) is ceased. Please use [travis-ci.com](http://www.travis-ci.com/?_gl=1*u8ascr*_ga*ODc4MTc4MDY5LjE2NDUxNjA0NDM.*_ga_XRYGSZFQ0P*MTY0NTQ4MDA1MC4yLjEuMTY0NTQ4MDA1OS41MQ..) from now on.

and [travis-ci.com's pricing](https://travis-ci.com/pricing/) clearly shows that there is no free plan, but only a free trial with the 10,000 credits I mentioned in my above post.

The reality is that Travis CI does not offer an indefinite free testing service anymore, as GitLab, GitHub Actions, and Azure Pipelines, and CircleCI (N.B. CircleCI is capped at 6,000 free build minutes per month) do.

**Comment by mbobra** [comment]
The JOSS [review criteria for automated tests](https://joss.readthedocs.io/en/latest/review_criteria.html#tests) only requires that authors "include an automated test suite covering the core functionality of their software." It does not require authors to run automated tests with any specific CI/CD platform.

**Comment by danehkar** [comment]
@matthewfeickert @transientlunatic 

OK, they changed it to limited 10,000 free credits. [ci.appveyor.com](https://ci.appveyor.com/) is still free for open source packages.

**Comment by henryiii** [comment]
Travis also now requires a credit card for new accounts. This caused a mess for a class I was auditing that had a CI lesson based on travis and everyone needed new accounts. (I got to teach the class impromptu over GHA instead, so that was fun :) )

If you email travis, you can sometimes get an open source monthly allowance. We do that for pypa projects. You need a very good reason to not use another system (GHA/Azure, CircleCI, GitLab CI, or maybe Appveyor, though Appveyor has gotten painfully slow). If you need specialized hardware, like ppc64le, Travis is good (though if you can use emulation, often still not the best).

PS: Have absolutely no context here. :)

**Comment by matthewfeickert** [comment]
(We should probably move any further discussion on CI and testing to a dedicated GitHub Issue :slightly_smiling_face:).

**Comment by danehkar** [comment]
@henryiii
Yes, Appveyor is very slow based on my experience. 

@transientlunatic 
You can test your code on a CI system of your choice (e.g. GitLab CI, GitHub Actions, Azure Pipelines, AppVeyor, and CircleCI).

**Comment by matthewfeickert** [comment]
@whedon commands

**Comment by matthewfeickert** [comment]
@editorialbot commands

**Comment by matthewfeickert** [comment]
@transientlunatic I want to apologize for not having finished opening Issues for the rest of the review. This last month has been a bit of a storm personally, but it isn't acceptable to have left you with no further feedback during this time. I will work to finish my full review before the start of next week.

**Comment by transientlunatic** [comment]
@matthewfeickert That's fine, I've been pretty busy myself, so I'm not complaining, and I think we've all been there one time or another, so please don't stress about it!

**Comment by mbobra** [comment]
@matthewfeickert No worries at all, please take your time :)

**Comment by mbobra** [comment]
👋 @danehkar @matthewfeickert How is it going with the review? Do you need any help? Please let me know!

**Comment by matthewfeickert** [comment]
> :wave: How is it going with the review? Do you need any help? Please let me know!

Thanks for the ping @mbobra! I should have my review finished this weekend (it is mostly there, just need to formalize it).

**Comment by mbobra** [comment]
👋 @danehkar Is everything going okay with this review? Please let me know if you need any help!

**Comment by danehkar** [comment]
@mbobra Thank you for the reminder. I will complete my review by the end of next week.

**Comment by danehkar** [comment]
@transientlunatic I checked the paper. 

- a typo in Line 15: Searches "Searches **of the** of the"
- While some dependent packages (e.g. htcondor) were mentioned, no mention of other packasges used in cli: [gwpy](https://gwpy.github.io/), [gwdatafind](https://git.ligo.org/computing/gwdatafind/client), [ligo-gracedb](https://gracedb.ligo.org/), [ligo-gracedb](https://git.ligo.org/lscsoft/ligo-segments). I am not sure if CLI is a part of the main package. However, pip and conda install these packages, which one installs ligo_asimov (pip) or ligo-asimov (conda).

**Comment by danehkar** [comment]
@transientlunatic Please add to [README.md](https://github.com/ligo-asimov/asimov#readme):

- the link to https://asimov.docs.ligo.org/asimov/master/index.html 
- Installation with Conda: `conda install -c conda-forge ligo-asimov`
- and fix issues with command lines in  README.md. e.g. `olivaw init `should be `olivaw init "Test project"`

**Comment by danehkar** [comment]
`olivaw init`
/anaconda3/envs/asimov/lib/python3.10/site-packages/htcondor/__init__.py:48: UserWarning: Neither the environment variable CONDOR_CONFIG, /etc/condor/, /usr/local/etc/, nor ~condor/ contain a condor_config source. Therefore, we are using a null condor_config.
  _warnings.warn(message)
Usage: olivaw init [OPTIONS] NAME
Try 'olivaw init --help' for help.
Error: Missing argument 'NAME'.

**Comment by danehkar** [comment]
`olivaw event create GW150914`
/anaconda3/envs/asimov/lib/python3.10/site-packages/htcondor/__init__.py:48: UserWarning: Neither the environment variable CONDOR_CONFIG, /etc/condor/, /usr/local/etc/, nor ~condor/ contain a condor_config source. Therefore, we are using a null condor_config.
  _warnings.warn(message)
Traceback (most recent call last):
  File "/anaconda3/envs/asimov/lib/python3.10/configparser.py", line 790, in get
    value = d[option]
  File "/anaconda3/envs/asimov/lib/python3.10/collections/__init__.py", line 986, in __getitem__
    return self.__missing__(key)            # support subclasses that define __missing__
  File "/anaconda3/envs/asimov/lib/python3.10/collections/__init__.py", line 978, in __missing__
    raise KeyError(key)
KeyError: 'engine'

**Comment by danehkar** [comment]
`olivaw production --event GW150914 --pipeline lalinference`
/anaconda3/envs/asimov/lib/python3.10/site-packages/htcondor/__init__.py:48: UserWarning: Neither the environment variable CONDOR_CONFIG, /etc/condor/, /usr/local/etc/, nor ~condor/ contain a condor_config source. Therefore, we are using a null condor_config.
  _warnings.warn(message)
Usage: olivaw production [OPTIONS] COMMAND [ARGS]...
Try 'olivaw production --help' for help.
Error: No such option: --event

**Comment by danehkar** [comment]
setup.py does not have all dependent packages. Installation with `conda install -c conda-forge ligo-asimov` was without any issues. However, I got the following error, when I installed with `pip install ligo-asimov` since the module **glue** was not installed.

olivaw init "Test project"

/anaconda3/lib/python3.9/site-packages/htcondor/__init__.py:49: UserWarning: Neither the environment variable CONDOR_CONFIG, /etc/condor/, /usr/local/etc/, nor ~condor/ contain a condor_config source. Therefore, we are using a null condor_config.
  _warnings.warn(message)
Traceback (most recent call last):
  File "/anaconda3/bin/olivaw", line 5, in <module>
    from asimov.olivaw import olivaw
  File "/anaconda3/lib/python3.9/site-packages/asimov/olivaw.py", line 26, in <module>
    from asimov.cli import event
  File "/anaconda3/lib/python3.9/site-packages/asimov/cli/event.py", line 12, in <module>
    from glue.lal import Cache
ModuleNotFoundError: No module named 'glue'

**Comment by danehkar** [comment]
@transientlunatic Based on packages are being installed when I run  `pip install ligo-asimov`, other packages also need to be listed in requirements.txt, which is used by `install_requires=requirements` in setup.py:
**gitdb
ligo-segments
ligotimegps
gwdatafind
requests-toolbelt
dqsegdb2
gwosc
safe-netrc
scitokens
mmap
cryptography
configparser**

as well as **glue**, you `import Cache from glue`, which is used by `datacache = Cache.from_urls(urls)` in event.py:
./asimov/cli/event.py:from glue.lal import Cache

**Comment by transientlunatic** [comment]
> @transientlunatic I checked the paper.
> 
>     * a typo in Line 15: Searches "Searches **of the** of the"
> 
>     * While some dependent packages (e.g. htcondor) were mentioned, no mention of other packasges used in cli: [gwpy](https://gwpy.github.io/), [gwdatafind](https://git.ligo.org/computing/gwdatafind/client), [ligo-gracedb](https://gracedb.ligo.org/), [ligo-gracedb](https://git.ligo.org/lscsoft/ligo-segments). I am not sure if CLI is a part of the main package. However, pip and conda install these packages, when one installs ligo_asimov (pip) or ligo-asimov (conda).

Thanks for catching the typo! 
I'm a bit confused about the second point; `gwdatafind` is a dependency of `gwpy` which is imported in asimov. Are you requesting that each dependency be mentioned in the paper?

**Comment by transientlunatic** [comment]
> @transientlunatic Based on packages are being installed when I run `pip install ligo-asimov`, other packages also need to be listed in requirements.txt, which is used by `install_requires=requirements` in setup.py: **gitdb ligo-segments ligotimegps gwdatafind requests-toolbelt dqsegdb2 gwosc safe-netrc scitokens mmap cryptography configparser**
> 
> as well as **glue**, you import Cache from glue (`from glue.lal import Cache`), which is used by `datacache = Cache.from_urls(urls)` in ./asimov/cli/event.py

Thanks for spotting these!
I don't believe any of these are requirements for asimov, but are requirements for some of its dependencies; I normally wouldn't list these as dependencies of the package itself, but if it's JOSS policy to do so I'm happy to add them.

**Comment by danehkar** [comment]
> > @transientlunatic I checked the paper.
> > ```
> > * a typo in Line 15: Searches "Searches **of the** of the"
> > 
> > * While some dependent packages (e.g. htcondor) were mentioned, no mention of other packasges used in cli: [gwpy](https://gwpy.github.io/), [gwdatafind](https://git.ligo.org/computing/gwdatafind/client), [ligo-gracedb](https://gracedb.ligo.org/), [ligo-gracedb](https://git.ligo.org/lscsoft/ligo-segments). I am not sure if CLI is a part of the main package. However, pip and conda install these packages, when one installs ligo_asimov (pip) or ligo-asimov (conda).
> > ```
> 
> Thanks for catching the typo! I'm a bit confused about the second point; `gwdatafind` is a dependency of `gwpy` which is imported in asimov. Are you requesting that each dependency be mentioned in the paper?

It is not necessary to mention all secondary dependencies in the paper, but the primary dependencies should be mentioned, for example [gwpy](https://github.com/gwpy/gwpy/), which was not mentioned in the paper:

```
./asimov/scripts/check-ifo.py:import gwpy
./asimov/scripts/check-ifo.py:import gwpy.timeseries
./asimov/scripts/check-ifo.py:import gwpy.segments
./asimov/scripts/check-ifo.py:from gwpy.segments import DataQualityFlag
./asimov/scripts/check-ifo.py:            state = gwpy.timeseries.StateVector.read(
./asimov/asimov/cli/production.py:import gwpy
./asimov/asimov/cli/event.py:import gwpy
./asimov/asimov/cli/event.py:import gwpy.timeseries
./asimov/asimov/cli/event.py:from gwpy.segments import DataQualityFlag
./asimov/asimov/cli/event.py:            state = gwpy.timeseries.StateVector.read(
```

**Comment by danehkar** [comment]
> Thanks for spotting these! I don't believe any of these are requirements for asimov, but are requirements for some of its dependencies; I normally wouldn't list these as dependencies of the package itself, but if it's JOSS policy to do so I'm happy to add them.

@transientlunatic There is an issue with the **pip** installation, but **conda** is OK. Those secondary dependencies are not required to be listed, but the primary dependencies should be in setup.py. If **glue** cannot be installed through **pip**, it should be mention in README.md, for example, users need to manually install the glue package for the `olivaw init` functions (e.g. using conda or etc).

**Comment by mbobra** [comment]
@transientlunatic How is it going? Are you finding the review process okay and do you need any help?

@matthewfeickert Is the review going okay? Are you waiting on the submitting author to make any changes?

**Comment by matthewfeickert** [comment]
> Is the review going okay? Are you waiting on the submitting author to make any changes?

Yes, sorry currently moving houses and need to find the time to finish the last bit. The review lag is entirely my fault, though @transientlunatic will need to close out [the Issues I have already](https://github.com/ligo-asimov/asimov/issues?q=is%3Aopen+is%3Aissue+author%3Amatthewfeickert+%5BJOSS+Review%5D) on https://github.com/ligo-asimov/asimov/issues and can get started on those whenever.

**Comment by transientlunatic** [comment]
> > Is the review going okay? Are you waiting on the submitting author to make any changes?
> 
> @mbobra Yes, sorry currently moving houses and need to find the time to finish the last bit. The review lag is entirely my fault, though @transientlunatic will need to close out [the Issues I have already](https://github.com/ligo-asimov/asimov/issues?q=is%3Aopen+is%3Aissue+author%3Amatthewfeickert+%5BJOSS+Review%5D) on https://github.com/ligo-asimov/asimov/issues and can get started on those whenever.

Yep, sorry for the radio silence; I'm currently advancing appropriate updates through internal LIGO review, but the Northern hemisphere summer vacation period has slowed the process a bit!

**Comment by mbobra** [comment]
@transientlunatic How are things going? Just wanted to check in.

**Comment by mbobra** [comment]
@transientlunatic Checking in again -- please let me know how things are going!

**Comment by mbobra** [comment]
@transientlunatic Checking in yet again -- please do let me know how things are going!

**Comment by transientlunatic** [comment]
> @transientlunatic Checking in yet again -- please do let me know how things are going!

I'm really sorry I've dropped the ball a bit on this. I think I've made most of the requested revisions for the reviewers, but I'll try and double check this and get things going again this week. Thanks for checking in on this and providing a bit of prodding!

**Comment by transientlunatic** [comment]
> > Is the review going okay? Are you waiting on the submitting author to make any changes?
> 
> @mbobra Yes, sorry currently moving houses and need to find the time to finish the last bit. The review lag is entirely my fault, though @transientlunatic will need to close out [the Issues I have already](https://github.com/ligo-asimov/asimov/issues?q=is%3Aopen+is%3Aissue+author%3Amatthewfeickert+%5BJOSS+Review%5D) on https://github.com/ligo-asimov/asimov/issues and can get started on those whenever.

Thanks @matthewfeickert I think I've now addressed all of the points you raised in issues on the repository (thank you for these!); I've closed them, and I hope that's okay, but please feel free to reopen if there's anything I've closed prematurely.

I'm just working through the comments from @danehkar again to make sure I've not missed anything.

Thanks both, and thanks @mbobra for keeping on top of this!

**Comment by mbobra** [comment]
No worries @transientlunatic. Happy to hear everything is going okay. Please update us when you've finished working through the remaining issues.

**Comment by mbobra** [comment]
👋 @transientlunatic Could you please give us a little update? No worries if you're still working on it!

**Comment by transientlunatic** [comment]
Ahh apologies @mbobra , this dropped off my radar again, and I still have a couple of comments to address.
I think I'm up to date with @matthewfeickert's though. 
I'll try and clear some time later this week to finish this.

**Comment by mbobra** [comment]
@transientlunatic Thank you!

**Comment by transientlunatic** [comment]
Hi @mbobra @matthewfeickert @danehkar 

I've now completed modifications to address all of the points raised in the review except for the issue regarding the need to specifically discuss all primary dependencies in the paper, and list all secondary dependencies (that is, dependencies of dependencies) in the package requirements. I'm not clear on the reasoning behind the latter I'm afraid, and I think an argument can be made that doing the former would make the paper much longer, given the relatively large number of primary dependencies, without adding clarity or useful information. I think, however, I may have misunderstood what is being asked for here, and I'm happy to add further text to address this point if I achieve greater understanding.

I'll also note that I've made one major update to the authorlist to reflect a name change of one author.

Thanks again for your patience with me getting this done!

**Comment by mbobra** [comment]
@danehkar @matthewfeickert Do you mind taking another look at the repository and paper when you have a moment? I'm sorry for keeping you on the line for a year here. Hopefully we're almost done.

In case it is helpful, here is the pointer the reviewer guidelines again: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. 

Please let me know if you need any help -- I know this has been a long journey, so I'm absolutely happy to help out.

**Comment by mbobra** [comment]
@danehkar @matthewfeickert Could you please take a quick look at the repository and give your final yay or nay? Also, if you don't have time to do this -- which I completely understand -- please let me know and I'll find another referee to finish the job. I'll make sure you both get credit for reviewing, regardless.

**Comment by mbobra** [comment]
@transientlunatic A small update from me -- I've contacted both referees by email.

**Comment by danehkar** [comment]
@mbobra @transientlunatic Thanks for your reminder. Where can I find the latest PDF file of the JOSS paper?

**Comment by matthewfeickert** [comment]
Oh wow I am very sorry. I do not know what happened but I have been getting none of these notifications for months. :disappointed: I'm not sure if something happened when I was migrating my GitHub notification system or not, but I'll look into it and try to get this done this week. Thank you for emailing me today @mbobra, else I don't think I would have seen this unless for chance.

**Comment by mbobra** [comment]
@editorialbot generate pdf from branch joss-paper

**Comment by mbobra** [comment]
@editorialbot generate pdf

**Comment by mbobra** [comment]
@editorialbot set joss-paper as branch

**Comment by mbobra** [comment]
@editorialbot generate pdf

**Comment by mbobra** [comment]
@danehkar Thanks for taking a look! You can find the most recent version of the paper directly above ⬆️

**Comment by danehkar** [comment]
@matthewfeickert My test with the package `asimov` installed using _pip_ was successful. However, `ligo-asimov` installed with _conda_ did not create  the `asimov` file in `bin` directory

_pip_:

     conda create --name asimov
     conda activate asimov
     pip install asimov
     simov init "my first project"
     ● New project created successfully!

_conda_:

     conda create --name asimov
     conda activate asimov
     conda install -c conda-forge ligo-asimov
     asimov init "Test project"         
     bash: asimov: command not found 

I finished my review. Please fix the issue of `ligo-asimov` with _conda_ in the future.

**Comment by danehkar** [comment]
@mbobra I do not have any additional comments. I did not see any issues with the package `asimov` , which was installed using _pip_, though there migth be issues if `ligo-asimov` is installed with _conda_. As my test with the _pip_ package `asimov` was successful, I would recommend `asimov` for publication. 

@matthewfeickert The API documentation can be improved if they are automatically created using [Sphinx](https://www.sphinx-doc.org) Python Documentation Generator using the [Sphinx RTD theme](https://pypi.org/project/sphinx-rtd-theme/).  The Sphinx creates the API documentation using the comments added after """ following each API declaration in the source code. They need to be specified  using `:Returns:`, `:Keywords:`, `:Examples:`, etc.

**Comment by matthewfeickert** [comment]
> However, `ligo-asimov` installed with _conda_ did not create the `asimov` file in `bin` directory
> 
> _pip_:
> 
> ```
>  conda create --name asimov
>  conda activate asimov
>  pip install asimov
>  simov init "my first project"
>  ● New project created successfully!
> ```
> 
> _conda_:
> 
> ```
>  conda create --name asimov
>  conda activate asimov
>  conda install -c conda-forge ligo-asimov
>  asimov init "Test project"         
>  bash: asimov: command not found 
> ```
> 
> I finished my review. Please fix the issue of `ligo-asimov` with _conda_ in the future.

@danehkar Can you make a GitHub Issue for this for them? https://github.com/ligo-asimov/asimov/issues

> The API documentation can be improved if they are automatically created using [Sphinx](https://www.sphinx-doc.org/) Python Documentation Generator using the [Sphinx RTD theme](https://pypi.org/project/sphinx-rtd-theme/). The Sphinx creates the API documentation using the comments added after """ following each API declaration in the source code. They need to be specified using :Returns:, :Keywords:, :Examples:, etc.

Yes, indeed. Sphinx is nice like this and I use it quite a bit for my packages. It isn't required of them to use any particular framework though, which is why I suggested this as an optional improvement.

**Comment by mbobra** [comment]
@danehkar @matthewfeickert Thank you so very much for your extreme patience and dedication throughout this process. I sincerely appreciate it. Your volunteer time keeps JOSS running and the JOSS team appreciates it very much ☀️

**Comment by mbobra** [comment]
@editorialbot check references

**Comment by mbobra** [comment]
Congratulations @transientlunatic -- we're almost ready to publish the paper! Could you please cut a new release of Asimov, [deposit it in Zenodo](https://guides.github.com/activities/citable-code/), and put the link to the Zenodo deposit in this thread? Please make sure that the author list and title on the Zenodo deposit matches the author list and title on the JOSS paper. Then I can move forward with accepting the paper.

**Comment by transientlunatic** [comment]
Thank you @mbobra!
I've deposited our latest release, 0.4.0, on zenodo, which has doi: [10.5281/zenodo.7843573](https://zenodo.org/record/7843573).

@danehkar, @matthewfeickert Thanks again for your patient and diligent work. I'll address the remaining points you raise as soon as possible; thanks again for these. I think the docs versioning is an easy fix at least; we do this already but it looks like a bit of javascript is broken, preventing you from seeing the menu!

**Comment by mbobra** [comment]
@editorialbot recommend-accept

**Comment by dfm** [comment]
@editorialbot accept

**Comment by dfm** [comment]
Many thanks to @danehkar and @matthewfeickert for reviewing, and to @mbobra for editing! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you!!

@transientlunatic — Your paper is now accepted and published in JOSS! :zap::rocket::boom:

**Comment by matthewfeickert** [comment]
Congratulations to @transientlunatic and coauthors! This was a very nice submission and it is excellent to see this software being used in the broader scientific community. :rocket:
